debian/mknfonts.1
.TH MKNFONTS 1 "March 12, 2004"
.SH NAME
mknfonts \- create nfont packages
.SH SYNOPSIS
.B mknfonts
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B mknfonts
command.
This manual page was written for the Debian distribution
because the original program does not have a manual page.
.PP
\fBmknfonts\fP doesn't completely automate the process of creating nfont-packages, but it can do most of the bulk work for you. Just put a bunch of fonts (.ttf, .pfa, .pfb, .afm, ...; most formats that freetype can handle and provide postscript names for) in an empty directory and run:
.TP
mknfonts *
.TP
You'll get a bunch of packages ready to be moved to .../Library/Fonts/. For best results, you'll want to review the packages. Most traits/weights (besides the really standard ones) won't be handled automatically, so you'll have to add or tweak them manually if you want GNUstep to know about them.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.SH SEE ALSO
.BR GNUstep (7).
.SH AUTHOR
This manual page was written by G\[:u]rkan Seng\[:u]n <gurkan@linuks.mine.nu>,
for the Debian project (but may be used by others).


debian/changelog
mknfonts.tool (0.5-11build2) oneiric; urgency=low

  * Rebuild with GCC 4.6.

 -- Matthias Klose <doko@ubuntu.com>  Sun, 14 Aug 2011 14:54:24 +0200

mknfonts.tool (0.5-11build1) natty; urgency=low

  * Rebuild for GNUstep transition.

 -- Luca Falavigna <dktrkranz@ubuntu.com>  Sun, 24 Oct 2010 14:43:16 +0200

mknfonts.tool (0.5-11) unstable; urgency=low

  * debian/control (Standards-Version): Compliant with 3.8.4 (no changes
    needed).
  * debian/rules (OPTFLAG): No longer define; rework noopt handling to be
    compatible with gnustep-make/2.4.x (Closes: #581962).
  * debian/source/format: New file.
  * debian/watch: Found it on ftp.gnustep.org; use that just in case.

 -- Yavor Doganov <yavor@gnu.org>  Sun, 30 May 2010 17:52:13 +0300

mknfonts.tool (0.5-10) unstable; urgency=low

  * debian/control (Maintainer): Adopt package; set to the GNUstep team
    (Closes: #454446).
    (Uploaders, Vcs-Arch): New fields.
    (Build-Depends): Drop gnustep-make.  Bump debhelper.  Remove version
    from libgnustep-base-dev.
    (Section): Change to `gnustep'.
    (Standards-Version): Set to 3.8.3 (no changes needed).
    (Provides, Replaces, Conflicts): Remove.
  * debian/compat: Set to 7.
  * debian/rules: Use $(MAKE) instead of gs_make.
    (LDFLAGS, OPTFLAG): Define.
    (build-stamp): Pass messages=yes, OPTFLAG and LDFLAGS to make.
    (install): Replace dh_clean -k with dh_prep.  Don't invoke
    dh_installdirs.
  * debian/dirs: Delete.
  * mknfonts.m (main): Don't delete the original font files; has a
    devastating effect on the system when used as root with packaged font
    files as arguments.
    (-[FaceInfo collectFaceInfo]): Use GSPrintf instead of fprintf.
  * GNUmakefile (ADDITIONAL_LDFLAGS): Rename as...
    (ADDITIONAL_TOOL_LIBS): ...to make --as-needed trim zlib as well.
  * debian/watch: New file (a stub).

 -- Yavor Doganov <yavor@gnu.org>  Thu, 03 Dec 2009 16:35:39 +0200

mknfonts.tool (0.5-9) unstable; urgency=medium

  * QA upload.
  * Fix bashism in /bin/sh script (Closes: #480608)
  * Migrate to Debhelper 5.
  * Tidy debian/rules.
  * debian/control:
    - Bump Standards-Version to 3.8.0
    - Add ${misc:Depends} on binary package 
    - Tidy long description

 -- Chris Lamb <chris@chris-lamb.co.uk>  Sat, 07 Jun 2008 22:51:10 +0100

mknfonts.tool (0.5-8) unstable; urgency=low

  * Orphaning package, setting maintainer to the Debian QA Group.

 -- Ana Beatriz Guerrero Lopez <ana@debian.org>  Fri, 07 Dec 2007 02:25:57 +0100

mknfonts.tool (0.5-7) unstable; urgency=low

  * GNUstep transition.

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Fri, 05 Oct 2007 00:04:33 +0200

mknfonts.tool (0.5-6) unstable; urgency=low

  * Rebuild against latest libgnustep-base-dev.
  * Bump standards version.

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Sat, 16 Sep 2006 12:17:31 +0200

mknfonts.tool (0.5-5) unstable; urgency=low

  * Rebuild against latest libgnustep-base. (Closes: #350079)
  * Update manual page, and move it to section 1.
  * Bump standards version.
  * Update update-nfonts.

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Mon, 23 Jan 2006 12:33:22 +0100

mknfonts.tool (0.5-4) unstable; urgency=low

  * Fixed a bug in update-nfonts (catch the last n<14 fonts).

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Wed,  8 Dec 2004 00:23:44 +0100

mknfonts.tool (0.5-3) unstable; urgency=low

  * Added update-nfonts script.
  * Renamed source package to mknfonts.tool.
  * Fixed debian/copyright.
  * Updated debian/control build-depends for GNUstep.

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Thu, 26 Aug 2004 00:34:45 +0200

mknfonts (0.5-2) unstable; urgency=low

  * Renamed binary package to mknfonts.tool
  * Updated for latest gnustep

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Fri, 30 Jul 2004 00:09:31 +0200

mknfonts (0.5-1) unstable; urgency=low

  * Initial Release.

 -- Gürkan Sengün <gurkan@linuks.mine.nu>  Fri, 12 Mar 2004 18:00:35 +0100


mknfonts.m
/*
copyright 2002 Alexander Malmberg <alexander@malmberg.org>
*/

#include <Foundation/NSString.h>
#include <Foundation/NSArray.h>
#include <Foundation/NSDictionary.h>
#include <Foundation/NSAutoreleasePool.h>
#include <Foundation/NSSet.h>
#include <Foundation/NSFileManager.h>

#include <ft2build.h>
#include FT_FREETYPE_H


static FT_Library ftlib;


/*
from GNUstep's core/back/Source/art/ftfont.m
GNU LGPL:ed
*/
#define NSBoldFontMask 0
#define NSItalicFontMask 0
#define NSCondensedFontMask 0
static int traits_from_string(NSString *s, unsigned int *traits, unsigned int *weight)
{
static struct
{
  NSString *str;
  unsigned int trait;
  int weight;
} suffix[] = {
/* TODO */
{@"Ultralight"     ,0                         , 1},
{@"Thin"           ,0                         , 2},
{@"Light"          ,0                         , 3},
{@"Extralight"     ,0                         , 3},
{@"Book"           ,0                         , 4},
{@"Regular"        ,0                         , 5},
{@"Plain"          ,0                         , 5},
{@"Display"        ,0                         , 5},
{@"Roman"          ,0                         , 5},
{@"Semilight"      ,0                         , 5},
{@"Medium"         ,0                         , 6},
{@"Demi"           ,0                         , 7},
{@"Demibold"       ,0                         , 7},
{@"Semi"           ,0                         , 8},
{@"Semibold"       ,0                         , 8},
{@"Bold"           ,NSBoldFontMask            , 9},
{@"Extra"          ,NSBoldFontMask            ,10},
{@"Extrabold"      ,NSBoldFontMask            ,10},
{@"Heavy"          ,NSBoldFontMask            ,11},
{@"Heavyface"      ,NSBoldFontMask            ,11},
{@"Ultrabold"      ,NSBoldFontMask            ,12},
{@"Black"          ,NSBoldFontMask            ,12},
{@"Ultra"          ,NSBoldFontMask            ,13},
{@"Ultrablack"     ,NSBoldFontMask            ,13},
{@"Fat"            ,NSBoldFontMask            ,13},
{@"Extrablack"     ,NSBoldFontMask            ,14},
{@"Obese"          ,NSBoldFontMask            ,14},
{@"Nord"           ,NSBoldFontMask            ,14},

{@"Italic"         ,NSItalicFontMask          ,-1},
{@"Oblique"        ,NSItalicFontMask          ,-1},

{@"Cond"           ,NSCondensedFontMask       ,-1},
{@"Condensed"      ,NSCondensedFontMask       ,-1},
{nil,0,-1}
};
  int i;

  *traits = 0;
//  printf("do '%@'\n", s);
  while ([s length] > 0)
    {
//      printf("  got '%@'\n", s);
      if ([s hasSuffix: @"-"] || [s hasSuffix: @" "])
	{
//	  printf("  do -\n");
	  s = [s substringToIndex: [s length] - 1];
	  continue;
	}
      for (i = 0; suffix[i].str; i++)
	{
	  if (![s hasSuffix: suffix[i].str])
	    continue;
//	  printf("  found '%@'\n", suffix[i].str);
	  if (suffix[i].weight != -1)
	    *weight = suffix[i].weight;
	  (*traits) |= suffix[i].trait;
	  s = [s substringToIndex: [s length] - [suffix[i].str length]];
	  break;
	}
      if (!suffix[i].str)
	break;
    }
//  printf("end up with '%@'\n", s);
  return [s length];
}



@interface FaceInfo : NSObject
{
@public
	NSString *postScriptName;
	NSString *familyName;
	NSString *faceName;

	NSMutableArray *files;
}
-(BOOL) collectFaceInfo;
-(NSDictionary *) faceInfoDictionary;
@end

@implementation FaceInfo
- init
{
	self=[super init];
	files=[[NSMutableArray alloc] init];
	return self;
}

-(void) dealloc
{
	DESTROY(files);
	[super dealloc];
}

-(NSString *) description
{
	return [NSString stringWithFormat: @"<FaceInfo: psname='%@' family='%@' face='%@' files=%@>",
		postScriptName,familyName,faceName,files];
}


-(BOOL) collectFaceInfo
{
	FT_Face face;
	int i;
	const char *psname;
	unsigned int foo,bar;

	for (i=1;i<[files count];i++)
	{
		NSString *f;
		f=[files objectAtIndex: i];
		if ([[f pathExtension] isEqualToString: @"pfa"] ||
			 [[f pathExtension] isEqualToString: @"pfb"])
		{
			[f retain];
			[files removeObject: f];
			[files insertObject: f atIndex: 0];
			[f release];
		}
	}

	if (FT_New_Face(ftlib,[[files objectAtIndex: 0] cString],0,&face))
	{
		GSPrintf(stderr,@"unable to open %@, ignoring\n",[files objectAtIndex: 0]);
		return NO;
	}
	for (i=1;i<[files count];i++)
	{
		if (FT_Attach_File(face,[[files objectAtIndex: i] cString]))
		{
			GSPrintf(stderr,@"warning: unable to attach %@\n",[files objectAtIndex: i]);
		}
	}

	psname=FT_Get_Postscript_Name(face);
	if (!psname)
	{
		GSPrintf(stderr,@"couldn't get postscript name for %@, ignoring\n",[files objectAtIndex: 0]);
		return NO;
	}
	postScriptName=[[NSString alloc] initWithCString: psname];

	familyName=[[NSString alloc] initWithCString: face->family_name];
	faceName=[[NSString alloc] initWithCString: face->style_name];

	if (traits_from_string(faceName,&foo,&bar))
	{
		GSPrintf(stderr,@"warning: couldn't fully parse '%@' (%@)\n",faceName,postScriptName);
	}

	FT_Done_Face(face);

	return YES;
}

-(NSDictionary *) faceInfoDictionary
{
	return [NSDictionary dictionaryWithObjectsAndKeys:
		files,@"Files",
		postScriptName,@"PostScriptName",
		faceName,@"Name",
		nil];
}

@end


static NSArray *group_files(NSArray *files)
{
	NSSet *combine_set=[NSSet setWithObjects: @"pfa",@"pfb",@"afm",@"pfm",nil];
	NSMutableDictionary *groups;
	NSArray *a;
	int i;

	groups=[[NSMutableDictionary alloc] init];
	for (i=0;i<[files count];i++)
	{
		NSString *f=[files objectAtIndex: i];
		NSString *key;
		FaceInfo *fa;

		if ([combine_set containsObject: [f pathExtension]])
			key=[f stringByDeletingPathExtension];
		else
			key=f;

		fa=[groups objectForKey: key];
		if (!fa)
		{
			fa=[[[FaceInfo alloc] init] autorelease];
			[groups setObject: fa  forKey: key];
		}
		[fa->files addObject: f];
	}

	a=[groups allValues];
	DESTROY(groups);
	return a;
}


int main(int argc, char **argv)
{
	CREATE_AUTORELEASE_POOL(arp);

	NSMutableArray *files;
	NSArray *groups;
	int i;

	NSMutableDictionary *families;


	files=[[NSMutableArray alloc] init];
	for (i=1;i<argc;i++)
	{
		[files addObject: [NSString stringWithCString: argv[i]]];
	}

	groups=[group_files(files) retain];
	DESTROY(files);

	if (FT_Init_FreeType(&ftlib))
	{
		fprintf(stderr,"failed to initialize freetype\n");
		return 1;
	}

	families=[[NSMutableDictionary alloc] init];
	for (i=0;i<[groups count];i++)
	{
		NSMutableArray *ma;
		FaceInfo *fi=[groups objectAtIndex: i];
		if ([fi collectFaceInfo])
		{
			ma=[families objectForKey: fi->familyName];
			if (!ma)
			{
				ma=[[[NSMutableArray alloc] init] autorelease];
				[families setObject: ma forKey: fi->familyName];
			}
			[ma addObject: fi];
		}
	}
	DESTROY(groups);

	{
		NSFileManager *fm=[NSFileManager defaultManager];
		NSEnumerator *e;
		NSString *family;
		int i,c;

		NSMutableDictionary *family_info;
		NSMutableArray *faces;
		NSArray *faceinfos;
		FaceInfo *fi;
		NSString *path;


		e=[families keyEnumerator];
		while ((family=[e nextObject]))
		{
			family_info=[[NSMutableDictionary alloc] init];
			faces=[[NSMutableArray alloc] init];
			[family_info setObject: faces  forKey: @"Faces"];

			path=[family stringByAppendingPathExtension: @"nfont"];

			[fm createDirectoryAtPath: path
				attributes: nil];

			faceinfos=[families objectForKey: family];
			for (c=[faceinfos count],i=0;i<c;i++)
			{
				fi=[faceinfos objectAtIndex: i];
				[faces addObject: [fi faceInfoDictionary]];
			}

			[family_info
				writeToFile: [path stringByAppendingPathComponent: @"FontInfo.plist"]
				atomically: NO];
			DESTROY(faces);
			DESTROY(family_info);
		}
	}
	DESTROY(families);

	DESTROY(arp);
	return 0;
}

